Upgraded authentication-api to 3.0.0-RC2

dmatej opened the pull request:
- to avoid copy and paste I moved loading the ServletContext to SecurityUtil
- generic maps were synced to implemented interfaces (usually `Map<Object, Object>` to `Map<String, Object>`)
- implemented new methods: registerServerAuthModule, removeServerAuthModule

Locally passed GlassFish tests including two which failed with 4.0.0-M3:
https://ci.eclipse.org/glassfish/job/glassfish_build-and-test-using-jenkinsfile/job/PR-23843/3/testReport/

dmatej committed:
Upgrade authentication-api.version from 2.0.0 to 3.0.0-RC2

dmatej commented on wsit/ws-sx/wssx-impl/src/main/java/com/sun/xml/wss/provider/wsit/JMACAuthConfigFactory.java:
```diff
@@ -235,24 +240,71 @@ public JMACAuthConfigFactory(ClassLoader loader) {
+
+
+    @Override
+    public String registerServerAuthModule(ServerAuthModule serverAuthModule, Object context) {
```
These two methods should be carefully reviewed; there is no test for them. The implementation is inspired by something similar what @arjantijms  made in Soteria, but customized for WSIT the case.

lukasj commented on wsit/ws-sx/wssx-impl/src/main/java/com/sun/xml/wss/impl/misc/SecurityUtil.java:
```diff
@@ -66,6 +67,9 @@
 
 import org.w3c.dom.Node;
+
+import static com.sun.xml.wss.provider.wsit.logging.LogStringsMessages.WSITPVD_0066_SERVLET_CONTEXT_NOTFOUND;
```
avoid static imports (there is currently none in the whole code base...)

dmatej replied:
Why are static imports an issue in 2022? They improve code readability since Java 5 (2004). I would rather expect that someone would remove all those trailing spaces, mixed tabs and spaces, TODOs and commented out code, fix broken formatting, cyclic dependencies between packages ... so much maintenance to do in many JEE projects ...

But if you really want it, you got it ... in the morning :-)

lukasj replied:
IMHO they make the code less readable

dmatej replied:
> (there is currently none in the whole code base...)

Actually this is not true, there are even static imports with asterisks, where I would agree it is not ok:
![image](https://user-images.githubusercontent.com/302904/157634296-b1c3db2b-ad72-42ab-88ec-2c61b063b18d.png)

lukasj replied:
right, don't know how the wrong part the comment got there. I swear it was deleted before the comment button was hit....

Anyway - all these got into the codebase with the comment "remove static imports" but as can be seen, people usually did not care and reason "someone else will do it" led to where we are (the same applies to tabs/spaces, formatting issues, etc....)

dmatej replied:
And another issue - in the SecurityUtil  is used several other LogStringMessages classes, one is in imports, other are used with full class name. Usually I don't need to know the full class name, I need just to be able to get to the message. Even Eclipse can go through using F3. Automatic optimization of imports doesn't break it too. So I would really prefer this

`log.log(Level.WARNING, WSITPVD_0066_SERVLET_CONTEXT_NOTFOUND());`

over
            
```
log.log(Level.WARNING,                
   com.sun.xml.wss.provider.wsit.logging.LogStringsMessages.WSITPVD_0066_SERVLET_CONTEXT_NOTFOUND());
```

lukasj replied:
all of those are coming from the wrong resource bundles in the first place. You moved the method but not the message it uses between resources bundles.

Correct thing to do is to move the message from `com.sun.xml.wss.provider.wsit.logging.LogStrings` to `com.sun.xml.wss.impl.misc.LogStrings` resource (in all language variations). If you do that for other two messages, no extra import is going to be required as the file will be in the same package

dmatej replied:
Do I understand well that you want me to move the message to com.sun.xml.wss.logging.impl.crypto.LogStringsMessages ?
What has it common with "crypto"?

But I have another idea - I can move those two methods from SecurityUtil to ReflectionUtil, it would be more appropriate place.

dmatej replied:
Seems even ReflectionUtil then could be renamed to ServletContextUtil or something similar ...

lukasj replied:
No.

the message with the key `WSITPVD_0066_SERVLET_CONTEXT_NOTFOUND` was used in `com.sun.xml.wss.provider.wsit.SecurityTubeFactory` (and it's the only usage I've found). After your change, this key is used in the `com.sun.xml.wss.impl.misc.SecurityUtil` and the resource bundle for this class is `com.sun.xml.wss.logging.impl.misc.LogStrings`, so the `WSITPVD_0066_SERVLET_CONTEXT_NOTFOUND` 
 message should be moved as well (`com.sun.xml.wss.logging.impl.misc.LogStringsMessages` is created from the resource file by the build)

lukasj replied:
basically move https://github.com/eclipse-ee4j/metro-wsit/blob/master/wsit/ws-sx/wssx-impl/src/main/resources/com/sun/xml/wss/provider/wsit/logging/LogStrings.properties#L130 to https://github.com/eclipse-ee4j/metro-wsit/blob/master/wsit/ws-sx/wssx-impl/src/main/resources/com/sun/xml/wss/logging/impl/misc/LogStrings.properties to follow the convention (I missed the "logging" part in the package name, sorry :-/ )

dmatej replied:
That will no help readability too, because LogStringMessages in imports lead to "crypto" package, not to "misc".
So let's do both (own class for SevletContext wrapping related things)  in next PR. I was already digging around, seems that the "zone" is quite small and can be in a single class, I can do that in an hour.

lukasj replied:
right, as you are in this, is moving https://github.com/eclipse-ee4j/metro-wsit/blob/master/wsit/ws-sx/wssx-impl/src/main/resources/com/sun/xml/wss/logging/impl/crypto/LogStrings.properties#L31 and https://github.com/eclipse-ee4j/metro-wsit/blob/master/wsit/ws-sx/wssx-impl/src/main/resources/com/sun/xml/wss/logging/LogStrings.properties#L1886 to the same resource file to avoid conflicts in imports not an option?

lukasj commented on wsit/ws-sx/wssx-impl/src/main/java/com/sun/xml/wss/impl/misc/SecurityUtil.java:
```diff
@@ -628,6 +632,49 @@ public static String getDataEncryptionAlgo(JAXBFilterProcessingContext context){
+     * @param endpoint
+     * @return null or the ServletContext instance bound to this endpoint
+     */
+    public static Object getServletContext(final WSEndpoint<?> endpoint) {
```
Why not to return `<T> T`?

dmatej replied:
Wouldn't it throw an exception if the ServletContext class would not be on the classpath? It expects an Object on all places, then sometimes even calls method via reflection.

lukasj replied:
see [type inference](https://docs.oracle.com/javase/tutorial/java/generics/genTypeInference.html) - in short as long as the type of the field the return value is assigned to remains Object, nothing changes; ideally, the servlet related code should be moved to some isolated place, since it is not required to be present (not something to be dealt with now).

dmatej replied:
This would not help anything because everywhere is expected java.lang.Object and the result object is always used as it is without retyping.
I agree it should be moved somewhere, but it must be a consistent work in own PR.

lukasj replied:
so it's better to leave it for the future instead of starting doing small steps towards the goal?

dmatej replied:
This is not a step towards goal. Goal would be some separate class/module managing ServletContext if it is available. Less reflection, no need for this, it would simply work directly with the ServletContext class. Until that it is better to keep consistency.

lukasj replied:
hm, are you saying that this consistency is good and consistency in avoiding static imports is bad?

dmatej replied:
> this consistency is good 

because there is no advantage of using generics here

> consistency in avoiding static imports is bad

because there is no such consistency as there is another 100 static imports in the repository

lukasj replied:
Are you going to fix those inconsistencies? If not then I would really appreciate you accept the fact that no new ones should be introduced. Thank you.

arjantijms replied:
> consistency in avoiding static imports is bad?

If they have never been avoided before, there is no new inconsistency right?

Also, why is not using static imports consistent? The project now also doesn't avoid imports. Isn't that not inconsistent too?

arjantijms replied:
@dmatej 
> I removed my static imports, and moved error messages.

Okay, it's a bit... questionable this entire request, but okay. @lukasj can this conversation now be resolved?

lukasj replied:
> Also, why is not using static imports consistent?

as I mentioned earlier - during a reviews back in 2006-7 it was noted to be removed and avoided. There was a promise they'll be removed but as usually - people come and go and promise does not hurt, right?

I really do not understand what it is so difficult to just respect the request to avoid using it (2 lines change done in 2minutes?) and spent endless hours arguing why it is "the best feature" added to the language instead.

arjantijms replied:
>  during a reviews back in 2006-7 it was noted to be removed and avoided.

Maybe we should have a review again, and this time decide they do not have to be removed, and do not have to be avoided?

In 15/16 years things have changed quite a bit. Maybe we don't have to avoid say lambdas now, because they were introduced after 2007. But if they were introduced in 2006, they would have to be avoided too because of that review?

arjantijms replied:
>  just respect the request to avoid using it 

Doesn't it go two ways? Maybe you could respect the request to not avoid them?

arjantijms replied:
But that all said, didn't @dmatej remove them already as mentioned?

> I removed my static imports, and moved error messages.

dmatej replied:
Honestly on my current experience - if you are the only one fighting against static imports, which is quite standard these days, while some people come to help, especially on the project in such painful state, it simply pushes any future help away. I don't need any response for this comment, just think about it.

lukasj replied:
I do appreciate the help and things are evolving and changing, yet - as a person who maintains this project for more than a decade, does all backports, patches etc and contributes to this project since its inception - I think I have the right to state the fact to the first time contributor(s) - especially when it's about a minute change. There are only two language features I really don't want to see - static imports (unless there is a really good reason for it - and there rarely is) and var.

dmatej commented:
Despite we differ in opinions I believe the https://github.com/eclipse-ee4j/metro-wsit/pull/162 is a compromise - I removed my static imports, and moved error messages. I found that despite SecurityUtil is in misc package, it is used mostly by the "crypto" parts, so I moved the remaining message there.
Things around ServletContext were moved to a wss own facade, which hides the optional dependency details. So no there is no need for reflection, generics, and we don't need to work with java.lang.Object except processing parameters.

lukasj commented:
obsolete
